### PR TITLE
Allow package exclusions and inclusions in javadocs

### PR DESCRIPTION
**Pull request by vinnybod**
There are times where developers want to exclude specific packages from being generated in the javadocs such as internal code. This is commonly offered as a feature of javadoc generation in both maven and gradle.

An exact package can be provided to remove the first level of files, or a `.*` to remove all subpackages under it.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -32,7 +33,9 @@ Generate a javadoc from all the `deps`
...
 | <a id="javadoc-doc_deps"></a>doc_deps |  `javadoc` targets referenced by the current target.<br><br>Use this to automatically add appropriate `-linkoffline` javadoc options to resolve references to packages documented by the given javadoc targets that have `url` specified.   | <a href="https://bazel.build/concepts/labels">List of labels</a> | optional |  `[]`  |
+| <a id="javadoc-doc_resources"></a>doc_resources |  Resources to include in the javadoc jar.   | <a href="https://bazel.build/concepts/labels">List of labels</a> | optional |  `[]`  |
 | <a id="javadoc-doc_url"></a>doc_url |  The URL at which this documentation will be hosted.<br><br>This information is only used by javadoc targets depending on this target.   | String | optional |  `""`  |
+| <a id="javadoc-excluded_packages"></a>excluded_packages |  A list of packages to exclude from the generated javadoc   | List of strings | optional |  `[]`  |
```
Comment (fzakaria) [inline docs/api.md:38]: nit: should include that it handles glob; and the glob is recursive with one *?

File: private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java [hunk 2]
```diff
@@ -175,7 +185,27 @@ public static void main(String[] args) throws IOException {
...
       options.addAll(Arrays.asList("-d", outputTo.toAbsolutePath().toString()));
 
-      sources.forEach(obj -> options.add(obj.getName()));
+      // sourcepath and subpackages should work in most cases.
+      // A known edge case is when the package names don't match the directory structure.
+      // For example `OneDep.java` in `tests/integration/maven_bom/OneDep.java` has a package
+      // of "com.github.bazelbuild.rules_jvm_external.example.maven_bom" but the file is in
+      // `tests/integration/maven_bom/OneDep.java`. I added the condition here to handle that case
+      // for now.
+      if (!excludedPackages.isEmpty()) {
```
Comment (fzakaria) [inline private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java:194]: you'll have to validate this but I think In java the rule is:
* source files don't have to match package directory
* class files DO have to match package directory

File: private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/BUILD [hunk 3]
```diff
@@ -1,6 +1,9 @@
 java_library(
     name = "create_jar_lib",
     srcs = ["CreateJar.java"],
+    visibility = [
+        "//visibility:public",
+    ],
```
[the remark this answers is not in the document]
Comment (vinnybod) [reply private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/BUILD:6]: Needs to be visible to the test below. I can scope it down if that's what you are asking

File: private/rules/kt_jvm_export.bzl [hunk 4]
```diff
@@ -88,8 +85,8 @@ def kt_jvm_export(
         name = name,
         maven_coordinates = maven_coordinates,
         lib_name = lib_name,
-        deploy_env = updated_deploy_env,
```
Comment (vinnybod) [inline private/rules/kt_jvm_export.bzl:91]: deploy_env here gets added back into the docs so we were generating docs for kotlin stdlib which seems wrong. Changed to excluded_workspace to keep it out of the docs. This is also what we do with scala already.

File: private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java [hunk 5]
```diff
@@ -248,16 +276,81 @@ private static void readSourceFiles(
...
+        if (line.startsWith("package ")) {
+          return line.substring("package ".length(), line.indexOf(';')).trim();
+        }
+
+        // Stop looking if we hit the class or interface declaration
+        if (line.startsWith("public")
+            || line.startsWith("class")
+            || line.startsWith("interface")
+            || line.startsWith("enum")) {
+          break;
```
Comment (fzakaria) [inline private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java:327]: optionally: Set.of("public", "class", "interface", "enum").contains(line)
(can also move the set outside)

**Review by fzakaria — APPROVED** [review]
LGTM -- we discussed offline some other approaches.
Perfect is the enemy of the good; I think we can improve if needed later and @vinnybod added a test so we can validate any change which is great.

We also discussed maybe a blacklist vs a whitelist with the pro/con being risks of accidentally including something we shouldn't.
I'm defering to Vince here for best judgement.

File: private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java (hunk 5 above)
Comment (vinnybod) [reply private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java:327]: Went with `keywords.stream().anyMatch(line::startsWith)`

**Review by fzakaria — APPROVED** [review]
